### PR TITLE
Feature/search

### PR DESCRIPTION
**Pull request by nathanajah**
Search in name, description, and person

### CODE REVIEW

File: multimedia/templates/multimedia/search_result.html [hunk 1]
```diff
@@ -0,0 +1,19 @@
+{% extends "multimedia/base_list.html" %}
+{% load staticfiles %}
+
+{% block title %}
+Applications - Digital Content Store
```
Comment (digawp) [inline multimedia/templates/multimedia/search_result.html]: _Eh-hem_

File: multimedia/views.py [hunk 2]
```diff
@@ -140,3 +140,16 @@ def application_detail(request, application_id):
...
+    kws = request.GET.get('keyword')
+    multimedia = Multimedia.objects.search(keywords=kws)
+
+    for item in multimedia:
+        item['url'] = '../' + item['url_prefix'] + '/' + item['url_suffix']
+        multimedia_image = MultimediaImage.objects.filter(multimedia_id=item['id']).first()
+        if multimedia_image:
+            item['thumbnail'] = multimedia_image.thumb150x150.url
+        else:
+            item['thumbnail'] = ''
```
Comment (digawp) [inline multimedia/views.py]: https://github.com/exclusive-gix/cs2102-store/blob/master/multimedia/views.py#L10
and
https://github.com/exclusive-gix/cs2102-store/blob/master/multimedia/views.py#L22

File: multimedia/managers.py [hunk 3]
```diff
@@ -157,3 +157,146 @@ def get(self, *args, **kwargs):
...
+    def search(self, multimedia_types=['application', 'movie', 'book', 'music'], **kwargs):
+        clauses = []
+        for mul_type in multimedia_types:
+            clauses.append(self.__get_whole_query(mul_type, **kwargs))
+        queries = []
+        for clause in clauses:
+            queries.append(clause[0])
+        substitutes = {}
+        for clause in clauses:
+            substitutes = self.__merge_dicts(substitutes, clause[1])
```
Comment (9gix) [inline multimedia/managers.py]: I prefer this line for simplicity

     substitutes.update(clasuses[1])

File: multimedia/templates/multimedia/search_result.html (hunk 1 above)
Comment (nathanajah) [inline multimedia/templates/multimedia/search_result.html]: Fixed.

File: multimedia/views.py (hunk 2 above)
Comment (nathanajah) [inline multimedia/views.py]: Fixed.

File: multimedia/views.py [hunk 4]
```diff
@@ -123,9 +123,7 @@ def application_list(request):
         multimedia_image = MultimediaImage.objects.filter(multimedia_id=application['id']).first()
         if multimedia_image:
             application['thumbnail'] = multimedia_image.thumb150x150.url
-        else:
-            application['thumbnail'] = no_preview_150
-
+        else: application['thumbnail'] = no_preview_150
```
Comment (digawp) [inline multimedia/views.py:135]: Why shift it into one line?

File: multimedia/managers.py [hunk 5]
```diff
@@ -157,3 +157,149 @@ def get(self, *args, **kwargs):
...
+
+        return (clause, substitutes)
+    def __merge_dicts(self, x, y):
+        z = x.copy()
+        z.update(y)
+        return z
+
+
+    def search(self, multimedia_types=['application', 'movie', 'book', 'music'], **kwargs):
+        print(multimedia_types)
```
Comment (digawp) [inline multimedia/managers.py]: Please clean up debugging codes

**Comment by digawp** [comment]
Ok this is some very meta thing you are doing here lol... Maybe some documentations on the private methods in the `MultimediaManager` would help

File: multimedia/managers.py [hunk 6]
```diff
@@ -157,3 +157,149 @@ def get(self, *args, **kwargs):
...
+        if len(multimedia_types) == 0:
+            multimedia_types = ['application', 'movie', 'book', 'music']
+        clauses = []
+        for mul_type in multimedia_types:
+            clauses.append(self.__get_whole_query(mul_type, **kwargs))
+        queries = []
+        for clause in clauses:
+            queries.append(clause[0])
```
Comment (digawp) [inline multimedia/managers.py]: Why do you need another copy of `clauses`?

**Comment by 9gix** [comment]
solve #24

**Comment by 9gix** [comment]
@nathanajah advance search for Book still not working

After tracing the query variable, This is the query generated when searching for 'algo' in 'book' type

```
SELECT
    m.id AS id,
    m.name AS name,
    m.description AS description,
    m.price AS price,
    'books' AS url_prefix,
    b.isbn13 AS url_suffix
FROM multimedia m, book b
WHERE m.id = b.multimedia_id
    AND (
        m.name LIKE '%algo%'
        OR m.description LIKE '%algo%'
        OR m.id IN (
            SELECT c.multimedia_id
            FROM crew c, person p
            WHERE c.person_id = p.id
                AND p.name LIKE '%algo%'
        )
    )
    AND m.id IN (
        SELECT m.multimedia_id
        FROM multimedia_category m, category c
        WHERE m.category_id = c.id
        AND c.name LIKE '%%'
    )
```

**Pull request merged**